### PR TITLE
Allow worker extensions to piggy-back on heartbeat

### PR DESCRIPTION
**Pull request by mrocklin**
To enable better diagnostics, it would be useful to allow worker
extensions to piggy-back on the standard heartbeat.  This adds an
optional "heartbeat" method to extensions, and, if present, calls a
custom method that gets sent to the scheduler and processed by an
extension of the same name.

This also starts to store the extensions on the worker in a named
dictionary.  Previously this was a list, but I'm not sure that it was
actually used anywhere.  This is a breaking change without deprecation,
but in a space that I suspect no one will care about.  I'm happy to
provide a fallback if desired.

### CODE REVIEW

**Comment by mrocklin** [comment]
This now also adds named extensions in the worker and scheduler

**Comment by mrocklin** [comment]
@fjetter are you ok with this change?  I'm changing the DEFAULT_EXTENSIONS collections from lists to named dictionaries.  Then secondarily I also add a mechanism to have extensions send messages up to their counterparts along with the regular heartbeat.

**Comment by mrocklin** [comment]
Merging in 24 hours if there are not further comments

**Review by fjetter — COMMENTED** [review]
I don't love how well this mechanism is hidden. For this to work, one needs to define a pair of extensions on both scheduler and worker side that share the same name.

So far, we also do not implement any hooks for extensions but only for plugins. I don't like the idea of introducing hooks for both extensions and plugins, particularly if we have things that are both an extension and a plugin (e.g. `stealing`)

**Comment by mrocklin** [comment]
> I don't love how well this mechanism is hidden. For this to work, one needs to define a pair of extensions on both scheduler and worker side that share the same name.  Correct.

> So far, we also do not implement any hooks for extensions but only for plugins. I don't like the idea of introducing hooks for both extensions and plugins, particularly if we have things that are both an extension and a plugin (e.g. stealing).

I agree that this is creating an implicit and hidden contract, which isn't great.  That being said, I don't currently see a much better option, other than making things very explicit and published (which seems premature to me, I don't yet have enough confidence that this is the one true way).  Thoughts?

File: distributed/tests/test_worker.py [hunk 1]
```diff
@@ -3306,3 +3306,38 @@ async def test_Worker__to_dict(c, s, a):
     }
     assert d["tasks"]["x"]["key"] == "x"
     assert d["data"] == ["x"]
+
+
+@gen_cluster(nthreads=[])
+async def test_extension_methods(s):
+    flag = [False]
+    shutdown = [False]
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/tests/test_worker.py:3314]: Done

**Comment by fjetter** [comment]
> That being said, I don't currently see a much better option, other than making things very explicit and published (which seems premature to me, I don't yet have enough confidence that this is the one true way). Thoughts?

Why do you want to use the heartbeat for this in the first place? We do have use cases for extensions to submit data regularly but haven't used the heartbeat before.
For instance, the `Semaphore` requires regular heartbeat-like messages to refresh leases, see https://github.com/dask/distributed/blob/4866615866af3daab115ea3da99fcc668a3e2482/distributed/semaphore.py#L432-L446
This originally used clients which was a scaling problem (many clients do not scale well right now) and we moved to reusing the worker comm pool in https://github.com/dask/distributed/pull/4195
We discussed using the `BatchedSend` instead but I decided against this for complexity reasons back then. However, if performance is of concern that may probably be the most straight forward approach.

The one property I see in the heartbeat that is _unique_ is its "scale interval to number of workers" property. Is this what you require?

**Comment by mrocklin** [comment]
> The one property I see in the heartbeat that is unique is its "scale interval to number of workers" property. Is this what you require?

Require is a strong word, but yes, that's the gesture that I'm trying to retain.  At the many hundred worker scale sending heartbeat like messages becomes a problem.  We should probably avoid it generally.  If semaphores are doing this then we should consider shifting them over to a system like this.  Are there other cases that you can think of?

**Comment by fjetter** [comment]
> Are there other cases that you can think of?

The client itself is not scaling well at all. I never debugged this properly but am suspecting heartbeats as one of the reasons. You have many clients when using tasks-in-tasks or other worker_client features. I've seen users being bothered by this without knowing it, e.g. when using prefect. I can also see the scheduler_info thing in the client... maybe more?

from the top of my head, I can't think of anything else. Briefly looking over the extensions, nothing jumps out at me

**Comment by mrocklin** [comment]
I'll try to take a look at the semaphore extension today and see if it makes sense to bring that over to something like this.  I imagine that this hasn't been as much of an issue just because semaphores are less commonly used.  For things like shuffle I would expect this to come up more often.  

(heads-up, I'm at a conference today and may be less responsive than normal)

**Comment by mrocklin** [comment]
I suggest that, as follow-up work, we consider replacing all extensions with plugins.  Thoughts?

**Comment by fjetter** [comment]
> I suggest that, as follow-up work, we consider replacing all extensions with plugins. Thoughts?

Sounds good. This PR should then use a plugin heartbeat and not an extension heartbeat, shouldn't it? This would also offer a natural place to document this

**Comment by mrocklin** [comment]
> Sounds good. This PR should then use a plugin heartbeat and not an extension heartbeat, shouldn't it? This would also offer a natural place to document this

It sounds like you're suggesting moving the heartbeat functionality from extensions to plugins.  Is that correct?  If so, would you also expect systems like shuffling to use plugins rather than extensions?  Semaphores too? 

In principle I think that it would be good to unify on one extension mechanism.  This would be a big change though.

**Comment by mrocklin** [comment]
> Sounds good. This PR should then use a plugin heartbeat and not an extension heartbeat, shouldn't it? This would also offer a natural place to document this

Actually, it looks like plugins don't any heartbeat mechanism today.  

I'm still thinking that this should go in, and then sometime in the future we should think about trying to unify plugins and heartbeats (and if that makes sense).  That seems like a distraction to me right now though.

**Comment by fjetter** [comment]
> It sounds like you're suggesting moving the heartbeat functionality from extensions to plugins. Is that correct? If so, would you also expect systems like shuffling to use plugins rather than extensions? 

This is how I understood your proposal. At this stage, I figured this would be an easy change. 

> Semaphores too?
> 
> In principle I think that it would be good to unify on one extension mechanism. This would be a big change though.

Moving everything is most definitely not in scope for this PR and I didn't want to suggest you doing all that work now.

> Actually, it looks like plugins don't any heartbeat mechanism today.

No it doesn't. nothing has a proper heartbeat mechanism and I figured this would be an easy way to make both sides happy.

**Comment by mrocklin** [comment]
Follow-up issue on extensions: https://github.com/dask/distributed/issues/6011

**Comment by mrocklin** [comment]
I spoke with @fjetter in a side conversation.  He said that he was comfortable enough with this approach given that there isn't a trivial approach to heartbeats elsewhere, and given that this is mostly just being used for diagnostics.  I plan to look over tests again and then merge.

**Pull request merged**